redshift/redshift_test.go

```diff
@@ -591,3 +465,92 @@ func TestTruncateInTimeRange(t *testing.T) {
 		t.Errorf("there were unfulfilled expections: %s", err)
 	}
 }
+
+func TestUpdateTable(t *testing.T) {
+	schema, table := "testschema", "tablename"
+
+	inputTable := Table{
+		Name: table,
+		// order incorrectly on purpose to ensure ordering works
+		Columns: []ColInfo{
+			{"test3", "boolean", "true", false, false, false, 0},
+			{"test2", "int", "100", true, false, true, 1},
+			{"id", "text", "", false, true, false, 0},
+			{"test4", "float", "false", false, false, false, 0},
+		},
+		Meta: Meta{Schema: schema},
+	}
+
+	db, mock, err := sqlmock.New()
+	assert.NoError(t, err)
+	defer db.Close()
+	mockRedshift := Redshift{db}
+
+	mock.ExpectBegin()
+	for _, updateSQL := range []string{
+		`ADD COLUMN id character varying(256) PRIMARY KEY `,
+		`ADD COLUMN test2 integer DEFAULT 100 NOT NULL SORTKEY DISTKEY `,
+		`ADD COLUMN test4 double precision`,
+	} {
+		sql := fmt.Sprintf(`ALTER TABLE "%s"."%s" (%s)`, schema, table, updateSQL)
+		regex := `ALTER TABLE ".*".".*" (.*)` // a little awk, but the prepare makes sure this is good
+
+		mock.ExpectPrepare(sql)
+		mock.ExpectExec(regex).WithArgs().WillReturnResult(sqlmock.NewResult(0, 0))
+	}
+	mock.ExpectCommit()
+
+	fewerColumnsTargetTable := Table{
+		Name: table,
+		Columns: []ColInfo{
+			{"test3", "boolean", "true", false, false, false, 0},
+		},
+		Meta: Meta{Schema: schema},
+	}
+	tx, err := mockRedshift.Begin()
+	assert.NoError(t, err)
+	assert.NoError(t, mockRedshift.UpdateTable(tx, fewerColumnsTargetTable, inputTable))
+	assert.NoError(t, tx.Commit())
+}
+
+func TestCheckSchemasSame(t *testing.T) {
+	t1 := Table{Columns: []ColInfo{
+		ColInfo{Name: "IntColumn", PrimaryKey: true},
+		ColInfo{Name: "DateColumn", PrimaryKey: true},
+	}}
+	t2 := t1
+	columnOps, errors := checkSchemas(t1, t2)
+	assert.Equal(t, 0, len(columnOps))
+	assert.Equal(t, 0, len(errors), fmt.Sprintf("errors: %+v\n", errors))
+}
+
+func TestCheckSchemasAddColumns(t *testing.T) {
+	t1 := Table{Columns: []ColInfo{
+		ColInfo{Name: "IntColumn", PrimaryKey: true},
+	}}
+	t2 := Table{Columns: []ColInfo{
+		ColInfo{Name: "IntColumn", PrimaryKey: true},
+		ColInfo{Name: "DateColumn", PrimaryKey: false},
+		ColInfo{Name: "IntColumn2", PrimaryKey: true},
+	}}
+	columnOps, errors := checkSchemas(t1, t2)
+	assert.Equal(t, 0, len(errors), fmt.Sprintf("errors: %+v\n", errors))
+	assert.Equal(t, 2, len(columnOps))
+}
+
+func TestCheckSchemasDiffs(t *testing.T) {
+	// Do a re-order and a type difference
```

Comment:
I've added some extra tests and check the number of errors returned. Also, with the current approach we can determine if string -> string errors occurred because we check that the names of the columns match.